### PR TITLE
Markdown rendering overhaul

### PR DESCRIPTION
**Pull request by andrew-boyarshin**
The majority of markdown files should now look great.
1. Implemented sidebar and footer (GitHub like). Header can also be done, but even GitHub doesn't implement one.
2. Implemented `[[...]]` image & link handling. It is quite a feature. It supports a variety of syntaxes.

``` markdown
![logo](http://libgdx.badlogicgames.com/img/logo.png)
[[[https://d3js.org/logo.svg|alt="D3js Logo"|height=210px|width=210px]]](https://d3js.org)
[[https://d3js.org/logo.svg|alt="D3js Logo"|height=210px|width=210px]]
[[Install]]
[[my internal link|internal-ref]]
![](000 hello world.png)
It is my great responsibility to send you... to http://google.com in order to find all your answers there.
The point is, <https://google.com> is more correct.
[Wiki](Home) ▸ **API Reference**
[[semantic versioning|http://semver.org]]
[[images/icon-install.png]]
```
1. Added the ability to display any kind of text blobs with rendering as Markdown. For example [guard](https://github.com/guard/guard/wiki) uses Textile for its Home wiki page. Textile and markdown have common roots, so that it renders perfectly with Markdown parser. Anyway, it is better to display something, than to display nothing.
2. Fixed a bug in markdown parser (stopped any special formatting after non-closeable tags like img)
3. I've mirrored about 15 wikis. Everything renders fine, all images are resolved perfectly (even html hardcoded should work in most cases).

### CODE REVIEW

File: cmd/web.go
Comment (metalmatze) [inline cmd/web.go:566]: Unrelated to the Markdown rendering?

File: modules/markdown/markdown.go
Comment (metalmatze) [inline modules/markdown/markdown.go]: Why would you commit all this? 😊

**Review by metalmatze — CHANGES_REQUESTED** [review]

File: cmd/web.go
Comment (andrew-boyarshin) [reply cmd/web.go:566]: @metalmatze Related closely. To support all types of embedding images into markdown files (both in wiki and repo itself) I changed wiki content loading scheme. Without this change markdown files with HTML `` tags inside would fail to load pictures.

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @metalmatze I can't quite catch the meaning. It is necessary for the system to work.

Comment (tboerger) [reply modules/markdown/markdown.go]: I think he's talking about the examples for the matches below.

File: cmd/web.go
Comment (metalmatze) [reply cmd/web.go:566]: I see, no problem. Just seemed unrelated to me at first glance.

File: modules/markdown/markdown.go
Comment (metalmatze) [reply modules/markdown/markdown.go]: Exactly. 😊

File: modules/markdown/markdown.go
Comment (bkcsoft) [inline modules/markdown/markdown.go]: Why are these exported?

File: routers/repo/wiki.go
Comment (bkcsoft) [inline routers/repo/wiki.go]: these two should not be exported?

**Review by bkcsoft — COMMENTED** [review]
`go fmt` and `go lint` after these

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: Missing `disabled` attribute on checkbox?

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: Remove all commented code

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: Isn't this necessary anymore? If not maybe we can delete them

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: Same as above

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: Global variables are bad. Maybe this can be a function argument?

File: public/js/index.js
Comment (andreynering) [inline public/js/index.js:616]: Any reason to move this from backend to frontend?

File: public/js/index.js
Comment (andreynering) [inline public/js/index.js:1392]: Same as above

File: routers/repo/wiki.go
Comment (andreynering) [inline routers/repo/wiki.go]: Don't prefix functions with `_`

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: Doesn't we need this anymore?

**Review by andreynering — COMMENTED** [review]
Thank you for the PR, this seems a very good addition. :+1: 

I just made some comments. Remove commented code, run `go fmt`, maybe add some test cases.

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @andreynering yes, to remove awful grey shadow over the checkbox I used `read-only` instead of `disabled`. In fact, it only influences appearance.

File: public/js/index.js
Comment (andrew-boyarshin) [reply public/js/index.js:616]: @andreynering I've experimented hard with it for a long time. Often rendering on backend resulted in broken children links or image links. So instead it is now done on frontend, with a better, conservative yet effective implementation, which operates the DOM, not the wall of text. No links or image urls are now broken, as now it respects tags and their children. It is quite fast, and it offloads servers.

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @bkcsoft will do!

**Comment by strk** [comment]
What's the milestone for this ?

**Comment by andrew-boyarshin** [comment]
Fixing all weak areas of PR, writing some additional tests, improving test coverage. I hope it will be done by weekends.

**Comment by andrew-boyarshin** [comment]
Ready to review once again. Coverage of markdown.go increased significantly, new tests cover more than half of the code (about 65%, compared to about 25% before the PR), greatly improved wiki page loading (rewritten system, optimizing and improving architecture and code itself). Made some changes to Markdown API, so that all systems work with changes. Fixed bugs found, now markdown files look even better.
It would be nice to get this into 1.0 release too!

File: modules/markdown/markdown.go
Comment (andreynering) [reply modules/markdown/markdown.go]: I think I prefer the old `disabled` here.

Also, the checkbox should also be disabled while previewing changes.

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: I found problems with issue pattern rendering.

Link `https://github.com/go-gitea/gitea/pull/186` renders to:
- Title: `#186`
- Preview URL: `http://localhost:3000/issues/186`
- After saved URL: `http://localhost:3000/andreynering/mylocalrepo/issues/186`

Link `https://github.com/go-gitea/gitea/pull/186/files` renders to:
- Title: `#186/files`
- Preview URL: `http://localhost:3000/issues/186`
- After saved URL: `http://localhost:3000/andreynering/mylocalrepo/issues/186`

The original URL should be preserved. It can refer to GitHub or another Gitea instance, as example.

File: modules/markdown/markdown.go
Comment (andreynering) [inline modules/markdown/markdown.go]: I noticed these kind of link always point to wiki. Is this intentional?

Example: `[[foobar]]` on `README.md` will refer to `/user/repo/wiki/foobar` and not a file called `foobar`.

File: modules/markdown/markdown_test.go
Comment (andreynering) [inline modules/markdown/markdown_test.go]:
``` go
oldAppUrl := setting.AppUrl
defer func() {
    setting.AppUrl = oldAppUrl
}()
setting.AppUrl = "http://gogs/"
```

**Review by andreynering — CHANGES_REQUESTED** [review]
There's still minor things that needs addressing.

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @andreynering yes, it is, it can be changed though. I've just never seen a single repo with markdown file in it using ```[[syntax]]```. Shall I implement it for repo too?

File: modules/markdown/markdown_test.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown_test.go]: Will do.

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @andreynering will change.
> Also, the checkbox should also be disabled while previewing changes.

Uhm... you mean markdown editor frontend? Or what?

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @andreynering will look into the case, not sure at the moment what to do with that.

File: modules/markdown/markdown.go
Comment (andrew-boyarshin) [reply modules/markdown/markdown.go]: @andreynering it is theme for another PR, isn't it? I'm not exactly sure how to do it at the moment, in fact, gogs/gitea frontend is quite an unknown land for me.

Comment (andreynering) [reply modules/markdown/markdown.go]: It's a bug, so should be fixed in this PR. I will take a look on what caused it when I have a little time. 👍 

For future contributions, I think it's a good idea to split big features in more PRs. It makes it easier and faster to review.

Let me know if you need help with this. We can push your branch to the main repo, so more people would have write access to it.

**Comment by lunny** [comment]
This need rebase and review continue!

**Comment by tboerger** [comment]
Still in conflicting state

**Comment by Bwko** [comment]
@STALKER2010 Any updates on this?

**Comment by lunny** [comment]
if nobody want to takeover this one, I will move it to v1.2

**Comment by lunny** [comment]
Please rebase.

**Comment by andrew-boyarshin** [comment]
@lunny In progress. Test system rework was quite a change.

**Comment by andrew-boyarshin** [comment]
Could someone explain why the build failed? I can only see lint suggestions, not in my code.
P.S. Looks like common PR issue.

**Comment by andrew-boyarshin** [comment]
Looks like Windows has done some strange stuff to file permissions. Fixing...
Wow. Git made such a mess...

**Comment by lunny** [comment]
Please rebase, it seems there are many unused changes.

**Comment by andrew-boyarshin** [comment]
@lunny Done. Had to recreate branch and export all changes to patch file due to merge conflicts in the past and Windows file permissions change recently. Also squashed all commits into one.
Anyone willing to review?

File: routers/api/v1/misc/markdown.go
Comment (andrew-boyarshin) [inline routers/api/v1/misc/markdown.go]: I'm not quite sure about this one. I can't recall doing that. Any thoughts? Is it good or bad?

**Comment by lunny** [comment]
Great. I will review this.

File: public/js/index.js
Comment (lunny) [reply public/js/index.js:616]: @STALKER2010 it's also frontend?

File: routers/api/v1/misc/markdown.go
Comment (lunny) [reply routers/api/v1/misc/markdown.go]: Maybe `repo.HTMLUrl()` ?

**Comment by lunny** [comment]
It seems this PR could not fix #652 ?

File: public/js/index.js
Comment (lunny) [reply public/js/index.js:616]: This is only preview, so put it in front-end?

Comment (andrew-boyarshin) [reply public/js/index.js:616]: @lunny (1st message) yes, as server-side autolinking often pulled apart tags, attributes. On client side we can manipulate DOM the way we want to, with no harm done to structure.
@tboerger about search engine indexing.
Default ```<http://google.com>``` markdown syntax is always done on server side. Autolinking in plain text is not defined by any markdown standard, so it's just nice stuff to have. If we do plain text autolinking on server, we can corrupt the whole document. Wrong thing to do, so I disabled it. If we do it on client, we may not linkify all possible links, but we do it in much safer way for the document. So, why not. Even if user has disabled JS in his browser, standard links will still work, plain text will not. Better than rendering corrupted document either way. Plus, most popular search engines (like Google, Bing, Yandex) execute JavaScript, so no harm done.

File: routers/api/v1/misc/markdown.go
Comment (andrew-boyarshin) [reply routers/api/v1/misc/markdown.go]: @lunny Hmm... Three different variants. Perhaps, invite more reviewers? Opinions matter.

**Comment by lunny** [comment]
and this #411

**Comment by andrew-boyarshin** [comment]
@lunny please, take a look at my comments.
It might be possible to fix #652, but I would prefer it as a separate PR. I don't want to make this PR any larger, and I would like this PR to be merged sooner.
#411 sample markdown file has multiple issues, so no wonder it is rendered incorrectly. It might be upstream issue (and most likely is), but that's not easy to fix without patching markdown parser.

File: public/js/index.js
Comment (andrew-boyarshin) [reply public/js/index.js:616]: @lunny (2nd question) no, as you see below, it is for both editor preview and final rendering. Reasoning is shown above.

**Review by andreynering — APPROVED** [review]

**Comment by andrew-boyarshin** [comment]
I've finally figured out the correct API code, it was different from both old code and my change. Fixed, added tests to the API (looks like noone ever did API tests in Gogs/Gitea before me). Also increased code coverage, added a couple of tests. Finally, it is completely ready (or at least I hope so).
@lunny ```LGTM``` is still in operation?

Uhm... Drone is unable to find new ```code.gitea.io/gitea/routers/api/v1/misc``` tests. Works fine for me. File is pushed. What is going on?

**Comment by andrew-boyarshin** [comment]
Anyone willing to review?

**Comment by strk** [comment]
I've tried but it's huge.
Do you have a test repository on try.gitea.io which we can clone
locally to see how your version changes things ?

**Comment by andrew-boyarshin** [comment]
@strk you mean good repos to test on? I test it on guard, libgdx, jquery, D3 wikis. There are also some screenshots in the original Gogs version of this PR.

**Comment by andrew-boyarshin** [comment]
There are also tests in this PR, which cover everything pretty good (if not to count API tests, which only work locally; strange thing, haven't figured out why it didn't work remotely).

**Review by metalmatze — APPROVED** [review]

**Comment by andrew-boyarshin** [comment]
OK, 1 out of 2 ```LGTM```s, two approved reviews. Final ```LGTM``` and good to go?

**Comment by andrew-boyarshin** [comment]
Anyone willing to review?

File: public/js/index.js
Comment (bkcsoft) [reply public/js/index.js:616]: > most popular search engines (like Google, Bing, Yandex) execute JavaScript, so no harm done.

They also infer a penalty for changing the context in JS 🙁

Comment (andrew-boyarshin) [reply public/js/index.js:616]: @bkcsoft the context? You mean, the web page content? I've looked it up, googled it and found no reference of any penalty. What I did find is that Google does not always run JS. Yet, I bet Google crawler is smart enough to do autolinking itself. And even if it isn't, the standard Markdown links should work flawlessly, we only lose unnecessary standard extensions. If anyone is willing to implement autolinking server-side after merging this PR, I would be glad to hear it.

**Comment by andreynering** [comment]
LGTM

I'll let the owners decide, but I'd merge right this after the v1.1 release, so people that use master may file any bugs that may happen until we have this in a stable release (v1.2).

**Comment by andreynering** [comment]
Let L-G-T-M work

**Comment by lunny** [comment]
@andreynering that's a good idea. We need a PR to stay on master for a long time to find more bugs. I'm thinking our release cycle. Maybe we should publish a release about 2 months? 1 month for merge new feature, one month for bug fixed. All the PRs on the second month should be merged at next release cycle.

**Comment by Lourens-Rich** [comment]
Sorry, i want to ask about my problem with markdown texts.
https://git.r208.tk/Lourens-Rich/Education/src/master/Readme.md
On this link we can see how a table looks uncorrected (Headers are misplaced). This worked normally before.

**Comment by andrew-boyarshin** [comment]
@Lourens-Rich it should be unrelated to this PR. Yet, I confirm the problem and the solution is to remove display: block from table, that is [_markdown.less#L243](https://github.com/go-gitea/gitea/blob/master/public/less/_markdown.less#L243). It solves the problem, although it makes table 100% wide in parent container (nothing can be to the left or right side of the table anyway).

File: modules/markdown/markdown.go
Comment (Morlinest) [inline modules/markdown/markdown.go:181]: @andrew-boyarshin You should remove "/p" tag only if you also remove starting tag (so only if u do append). This code cause bugs in rendering (example: a list rendering bug with code in list - "/li" is replaced by "/p" in postprocess because the last starting tag before "pre, code" is only starting "p" tag).

**Pull request merged**
